extensions/amp-video/1.0/video-wrapper.js

```diff
@@ -0,0 +1,210 @@
+
+  const [muted, setMuted] = useState(autoplay);
+  const [metadata, setMetadata] = useState(null);
+  const [playing, setPlaying] = useState(false);
+  const [userInteracted, setUserInteracted] = useState(false);
+
+  const wrapperRef = useRef();
+  const playerRef = useRef();
+
+  const readyDeferredRef = useRef(new Deferred());
+
+  const play = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.play();
+    });
+  };
+
+  const pause = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.pause();
+    });
+  };
+
+  useEffect(() => {
+    if (!mediasession || !metadata || !playing) {
+      return;
+    }
+    setMediaSession(window, metadata, play, pause);
+  }, [mediasession, playing, metadata]);
+
+  return (
+    <div
+      ref={wrapperRef}
+      style={{position: 'relative', width: '100%', height: '100%'}}
+    >
+      <Component
+        ref={playerRef}
+        // Ensure that props controlled here are excluded from passthroughProps,
+        // see destructuring statement above.
+        muted={muted}
+        controls={controls && (!autoplay || userInteracted)}
+        onLoad={() => readyDeferredRef.current.resolve()}
+        onLoadedMetadata={() => {
+          if (!mediasession) {
+            return;
+          }
+
+          const metadata = playerRef.current.getMetadata() || {
+            ...EMPTY_METADATA,
+            artwork: undefined,
+          };
+
+          metadata.title =
+            metadata.title ||
+            passthroughProps.title ||
+            passthroughProps['aria-label'] ||
+            document.title;
+
+          metadata.artwork =
+            metadata.artwork ||
+            parseSchemaImage(document) ||
+            parseOgImage(document) ||
+            parseFavicon(document);
+
+          setMetadata(metadata);
+        }}
+        onPlaying={() => setPlaying(true)}
+        onPause={() => setPlaying(false)}
+        {...passthroughProps}
+      >
+        {children}
+      </Component>
+      {autoplay && !userInteracted && (
+        <Autoplay
+          playing={playing}
+          displayIcon={!noaudio && muted}
+          wrapperRef={wrapperRef}
+          play={play}
+          pause={pause}
+          displayOverlay={controls}
+          onOverlayClick={() => {
+            setMuted(false);
+            setUserInteracted(true);
+          }}
+        />
+      )}
+    </div>
+  );
+}
+
+/**
+ * @param {!VideoAutoplayProps} props
+ * @return {PreactDef.Renderable}
+ */
+function Autoplay({
+  displayIcon,
+  playing,
+  displayOverlay,
+  onOverlayClick,
+  wrapperRef,
+  play,
+  pause,
+}) {
+  useEffect(() => {
+    let observer = new IntersectionObserver(
+      (entries) => {
+        if (entries[0].isIntersecting) {
+          play();
+        } else {
+          pause();
+        }
+      },
+      {threshold: MIN_VISIBILITY_RATIO_FOR_AUTOPLAY}
+    );
+
+    observer.observe(wrapperRef.current);
```

Comment:
Similarly to above, this will run in order in a way that `wrapperRef` will be available. Otherwise, it's breaking the internal contract and escaping when this is nullish can cause us to miss resulting errors.

Reply:
Yeah, I think it's ok to require a `Ref` value in some cases. Though this really depends. Since this is a component called downstream, I think it's generally ok.